src/client/array/dc_array.c

```diff
@@ -1080,6 +1080,91 @@ create_sgl(d_sg_list_t *user_sgl, daos_size_t cell_size,
 	return 0;
 }
 
+static int
+set_short_read_cb(tse_task_t *task, void *data)
+{
+	daos_array_io_t		*args = daos_task_get_args(task);
+	struct io_params	*io_list = *((struct io_params **)data);
+	struct io_params	*current;
+	uint64_t		dkey_val;
+	daos_size_t		num_records = 0;
+	bool			break_on_lower = false;
+	int			rc = task->dt_result;
+
+	if (rc != 0) {
+		D_ERROR("Array Read Failed (%d)\n", rc);
+		return rc;
+	}
+
+	/*
+	 * List is already sorted in decreasing dkey order, so we just have to
+	 * look at the highest dkey with valid data.
+	 */
+	dkey_val = io_list->dkey_val;
+	current = io_list;
+	while (current) {
+		daos_size_t len = 0, recs;
+		int i;
+
+		/*
+		 * if we moved to a lower dkey and the higher one is not empty
+		 * or all short-fetched, we can break here.
+		 */
+		if (break_on_lower && dkey_val > current->dkey_val)
+			break;
+
+		/** if the sgl is empty then skip this entry */
+		if (current->sgl.sg_nr == 0)
+			goto next;
+
+		dkey_val = current->dkey_val;
+
+		/*
+		 * if no DAOS "short-fetch" detected, continue. Can't break here
+		 * because we could have the same dkey in the next entry that we
+		 * need to check.
+		 */
+		i = current->sgl.sg_nr_out - 1;
```

Comment:
current->sgl.sg_nr_out is possible to get 0 (for example read empty file)? in that case the below code will be segfault.

Reply:
no that is not possible, bec if nr_out is 0, next 2 checks will be skipped and we jump to line 1142 in patch.

Reply:
Ah right, thx